File: Modules/benefits-admin/Services/BenefitsService.cs
// Modules/benefits-admin/Services/BenefitsService.cs
using Grpc.Core;
using BenefitsAdministration.Protos;

namespace BenefitsAdministration.Services;

public class BenefitsService : Protos.BenefitsService.BenefitsServiceBase
{
    private readonly TaxRateService _taxRateService;
    private readonly ILogger<BenefitsService> _logger;

    public BenefitsService(TaxRateService taxRateService, ILogger<BenefitsService> logger)
    {
        _taxRateService = taxRateService;
        _logger = logger;
    }

    public override Task<BenefitCalculationResponse> CalculateBenefits(BenefitCalculationRequest request, ServerCallContext context)
    {
        _logger.LogInformation("Calculating benefits for claim {ClaimId}", request.ClaimId);
        
        // Calculate total wages and weeks worked
        double totalWages = 0;
        int totalWeeks = 0;
        
        foreach (var employment in request.EmploymentHistory)
        {
            // Calculate weeks worked
            var startDate = DateTime.Parse(employment.StartDate);
            var endDate = DateTime.Parse(employment.EndDate);
            var daysWorked = (endDate - startDate).TotalDays;
            var weeksWorked = (int)Math.Round(daysWorked / 7);
            
            // Calculate weekly wage based on pay frequency
            double weeklyWage = 0;
            
            switch (employment.PayFrequency.ToLower())
            {
                case "hourly":
                    weeklyWage = employment.WageRate * (employment.HoursPerWeek > 0 ? employment.HoursPerWeek : 40);
                    break;
                case "weekly":
                    weeklyWage = employment.WageRate;
                    break;
                case "biweekly":
                    weeklyWage = employment.WageRate / 2;
                    break;
                case "monthly":
                    weeklyWage = (employment.WageRate * 12) / 52;
                    break;
                case "annually":
                    weeklyWage = employment.WageRate / 52;
                    break;
                default:
                    weeklyWage = 0;
                    break;
            }
            
            totalWages += weeklyWage * weeksWorked;
            totalWeeks += weeksWorked;
        }
        
        // Calculate average weekly wage
        double averageWeeklyWage = totalWeeks > 0 ? totalWages / totalWeeks : 0;
        
        // Weekly benefit amount (typically 50% of average weekly wage, capped)
        double weeklyBenefitAmount = Math.Min(averageWeeklyWage * 0.5, 500);
        
        // Maximum benefit amount (typically 26 weeks of benefits)
        double maximumBenefitAmount = weeklyBenefitAmount * 26;
        
        // Get tax rate
        double taxRate = _taxRateService.GetTaxRate("US-DEFAULT");
        
        // Calculate net weekly benefit amount
        double netWeeklyBenefitAmount = weeklyBenefitAmount * (1 - taxRate);
        
        // Set benefit year
        var now = DateTime.UtcNow;
        var benefitYearEnd = now.AddYears(1);
        
        return Task.FromResult(new BenefitCalculationResponse
        {
            ClaimId = request.ClaimId,
            WeeklyBenefitAmount = weeklyBenefitAmount,
            MaximumBenefitAmount = maximumBenefitAmount,
            BenefitYear = new BenefitYear
            {
                StartDate = now.ToString("yyyy-MM-dd"),
                EndDate = benefitYearEnd.ToString("yyyy-MM-dd")
            },
            TaxWithholdingRate = taxRate,
            NetWeeklyBenefitAmount = netWeeklyBenefitAmount
        });
    }

    public override Task<TaxRateResponse> GetTaxWithholdingRate(TaxRateRequest request, ServerCallContext context)
    {
        _logger.LogInformation("Getting tax rate for state {StateCode}", request.StateCode);
        
        var taxRate = _taxRateService.GetTaxRate(request.StateCode);
        
        return Task.FromResult(new TaxRateResponse
        {
            TaxWithholdingRate = taxRate
        });
    }

    public override Task<TaxRateResponse> UpdateTaxWithholdingRate(UpdateTaxRateRequest request, ServerCallContext context)
    {
        _logger.LogInformation("Updating tax rate for state {StateCode} to {NewRate}", request.StateCode, request.TaxWithholdingRate);
        
        _taxRateService.UpdateTaxRate(request.StateCode, request.TaxWithholdingRate);
        
        return Task.FromResult(new TaxRateResponse
        {
            TaxWithholdingRate = request.TaxWithholdingRate
        });
    }
}

File: modules/claims-processing/src/services/benefitsService.js
// Modules/claims-processing/src/services/benefitsService.js
const axios = require('axios');

const API_GATEWAY_URL = process.env.API_GATEWAY_URL || 'http://localhost:8000';
const BENEFITS_SERVICE_URL = `${API_GATEWAY_URL}/api/benefits`;

// Calculate benefits for a claim
exports.calculateBenefits = async (benefitRequest) => {
  try {
    // For demo purposes, we'll mock the benefit calculation
    // In a real implementation, you would call the benefits service
    // const response = await axios.post(`${BENEFITS_SERVICE_URL}/calculate`, benefitRequest);
    // return response.data;
    
    console.log('Benefits calculation request:', benefitRequest);
    
    // Simulate a basic benefit calculation
    let totalWages = 0;
    let totalWeeks = 0;
    
    for (const employment of benefitRequest.employmentHistory) {
      // Calculate total wages based on employment details
      const startDate = new Date(employment.startDate);
      const endDate = new Date(employment.endDate);
      
      // Calculate weeks worked (simplified)
      const millisecondsPerWeek = 7 * 24 * 60 * 60 * 1000;
      const weeksWorked = Math.round((endDate - startDate) / millisecondsPerWeek);
      
      // Calculate wages based on pay frequency
      let weeklyWage = 0;
      
      switch (employment.payFrequency) {
        case 'Hourly':
          weeklyWage = employment.wageRate * (employment.hoursPerWeek || 40);
          break;
        case 'Weekly':
          weeklyWage = employment.wageRate;
          break;
        case 'Biweekly':
          weeklyWage = employment.wageRate / 2;
          break;
        case 'Monthly':
          weeklyWage = (employment.wageRate * 12) / 52;
          break;
        case 'Annually':
          weeklyWage = employment.wageRate / 52;
          break;
        default:
          weeklyWage = 0;
      }
      
      totalWages += weeklyWage * weeksWorked;
      totalWeeks += weeksWorked;
    }
    
    // Average weekly wage
    const averageWeeklyWage = totalWeeks > 0 ? totalWages / totalWeeks : 0;
    
    // Weekly benefit amount (typically 50% of average weekly wage, capped)
    const weeklyBenefitAmount = Math.min(averageWeeklyWage * 0.5, 500);
    
    // Maximum benefit amount (typically 26 weeks of benefits)
    const maximumBenefitAmount = weeklyBenefitAmount * 26;
    
    // Tax withholding (mock rate of 10%)
    const taxWithholdingRate = 0.1;
    const netWeeklyBenefitAmount = weeklyBenefitAmount * (1 - taxWithholdingRate);
    
    // Benefit year
    const now = new Date();
    const benefitYearStart = now.toISOString();
    const benefitYearEnd = new Date(now.setFullYear(now.getFullYear() + 1)).toISOString();
    
    return {
      claimId: benefitRequest.claimId,
      weeklyBenefitAmount,
      maximumBenefitAmount,
      benefitYear: {
        startDate: benefitYearStart,
        endDate: benefitYearEnd
      },
      taxWithholdingRate,
      netWeeklyBenefitAmount
    };
  } catch (error) {
    console.error('Error calculating benefits:', error);
    throw new Error('Failed to calculate benefits');
  }
};
